src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,724 @@
+}
+
+/// `RegionEventListener` implements observer traits. It simply send the events that we are interested in
+/// through the `scheduler`.
+#[derive(Clone)]
+struct RegionEventListener {
+    scheduler: Scheduler<RegionCollectionMsg>,
+}
+
+impl Coprocessor for RegionEventListener {}
+
+impl RegionChangeObserver for RegionEventListener {
+    fn on_region_changed(&self, context: &mut ObserverContext, event: RegionChangeEvent) {
+        let region = context.region().clone();
+        let event = match event {
+            RegionChangeEvent::Create => RaftStoreEvent::CreateRegion { region },
+            RegionChangeEvent::Update => RaftStoreEvent::UpdateRegion { region },
+            RegionChangeEvent::Destroy => RaftStoreEvent::DestroyRegion { region },
+        };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+impl RoleObserver for RegionEventListener {
+    fn on_role_change(&self, context: &mut ObserverContext, role: StateRole) {
+        let region = context.region().clone();
+        let event = RaftStoreEvent::RoleChange { region, role };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+/// Create an `RegionEventListener` and register it to given coprocessor host.
+fn register_region_event_listener(
+    host: &mut CoprocessorHost,
+    scheduler: Scheduler<RegionCollectionMsg>,
+) {
+    let listener = RegionEventListener { scheduler };
+
+    host.registry
+        .register_role_observer(1, box listener.clone());
+    host.registry
+        .register_region_change_observer(1, box listener);
+}
+
+/// `RegionCollection` is the place where we hold all region information we collected, and the
+/// underlying runner of `RegionInfoAccessor`. It listens on events sent by the `RegionEventListener` and
+/// keeps information of all regions. Role of each region are also tracked.
+struct RegionCollection {
+    // HashMap: region_id -> (Region, State)
+    regions: RegionsMap,
+    // BTreeMap: 'z' + end_key -> region_id
+    region_ranges: RegionRangesMap,
+}
+
+impl RegionCollection {
+    fn new() -> Self {
+        Self {
+            regions: HashMap::default(),
+            region_ranges: BTreeMap::default(),
+        }
+    }
+
+    /// Check if the end_key has collided with another region. If so, try to clean it.
+    /// If the collided region has a greater version than the current one, it means that the current
+    /// operation is caused by a stale message. Then this function should fail.
+    ///
+    /// Returns a bool value indicating whether checking succeeded.
+    fn check_end_key_collision(&mut self, region_id: u64, version: u64, end_key: &[u8]) -> bool {
+        if let Some(collided_region_id) = self.region_ranges.get(end_key).cloned() {
+            // There is already another region with the same end_key
+            assert_ne!(collided_region_id, region_id);
```

Comment:
Is it possible that the same message arrive twice?

Reply:
This function is called only when creating a region or updating a region's end_key. The second creation will be regarded as an update, and I'll add some code to ignore stale messages.